PixelPaint: Ensure layers are painted consistently when panning and zooming

tcl3 opened the pull request:
This patch aims to fix an issue where whitespace would appear at the edges of a layer when panning and zooming.

This behaviour was caused by rounding errors when calculating the position and size of the destination rect in `Image::paint_into()`.

Here is a video showing the edge of the canvas appearing and disappearing at the bottom and right edge of the layer when zooming:
 (you may need to enlarge the video to see the offending whitespace.)

https://user-images.githubusercontent.com/2817754/210777858-51bc821d-d5e4-4255-a764-62227b13c3fb.mp4

After:

https://user-images.githubusercontent.com/2817754/210777903-de123677-6e71-4d6e-b18c-954cfb3d1b2c.mp4

GMTA commented on Userland/Applications/PixelPaint/ImageEditor.cpp:
```diff
@@ -167,7 +167,7 @@ void ImageEditor::paint_event(GUI::PaintEvent& event)
 
     if (m_active_layer && m_show_active_layer_boundary)
-        painter.draw_rect(content_to_frame_rect(m_active_layer->relative_rect()).to_rounded<int>().inflated(2, 2), Color::Black);
+        painter.draw_rect(content_to_frame_rect(m_active_layer->relative_rect()).to_type<int>().inflated(2, 2), Color::Black);
```
This arguably gives you a worse result since instead of rounding, you're now flooring the value. Isn't the issue that there is a cast to int somewhere else?

tcl3 replied:
Everywhere else in `ImageEditor` uses truncation when converting float coordinates to int. Either via `to_type<int>()` or implicitly in methods like `frame_to_content_position()`, `frame_to_content_rect()` and others. The idea of this change was to make layer boundary drawing consistent with everything else.

tcl3 replied:
@GMTA - I've just pushed an alternative implementation, that fixes `Image::paint_into()` directly. I believe this better shows the issues that led to the problem.

I have also squashed together the commits that fix layer painting and layer boundary painting because, as you say, they each give a worse result when applied individually, but are both necessary for a correct result.

tcl3 commented:
@GMTA - I've just pushed an alternative implementation, that fixes Image::paint_into() directly. I believe this better shows the issues that led to the problem.

I have also squashed together the commits that fix layer painting and layer boundary painting because, as you say, they each give a worse result when applied individually, but are both necessary for a correct result.

tcl3 commented:
@GMTA - I've rebased  and added a small change, that draws the layer boundary over the pixel grid

I've done more testing and, from what I can see, the layer content now lines up with the layer boundary and the canvas boundary. However, there is a remaining issue with the alignment of the pixel grid. 

I believe this is caused by the fact that we round twice when calculating each line co-ordinate for the grid. Can you think of a simple way to minimise the rounding error when calculating the line coordinates?

GMTA commented:
@tcl3 What issue exactly? Do you have an example?

tcl3 commented:
Sorry, I've just reread what I wrote and I didn't actually explain the problem. 

Here's a video that shows the issue with  the pixel grid:

https://user-images.githubusercontent.com/2817754/213307262-2deffd0a-cd92-4165-a0ec-2e91fafc5104.mp4

The edges of the pixel grid next to the layer boundary appear and disappear as we zoom in and out. This is because the grid is moving back and forth by one pixel relative to the layer and the layer boundary.

_I think_ the cause of this is a rounding error when calculating the start and end points of pixel grid lines.

tcl3 commented:
Latest push removes the second commit, changing when the layer boundary is drawn. I have a better solution to ensure the layer boundary is always visible, which I'm going to open a separate PR for. 

I believe this should be ready to merge now :^)
